### DISCUSSION TITLE
Buffer size for file transfer client server application

### DISCUSSION THREAD

**Discussion by kamranattari4192** [microsoft/msquic#3679] [Q&A]
Hi,

For a file transfer client server application with MsQuic, what should be the optimal buffer size on sending and receiving side?

**Comment by nibanks** [comment]
The sizes generally depend on the scenario and the network conditions. The `QUIC_STREAM_EVENT_IDEAL_SEND_BUFFER_SIZE` event indicates how much you should keep pending on the send side to ensure you don't limit throughput. On the receive side, just process the data as fast as you can and MsQuic automatically matches the flow control buffer for you.

**Reply by janksenhu** [reply]
i use https://github.com/microsoft/msquic/blob/main/src/tools/sample/sample.c

send 1GB file on ubuntu 16.04

find the server side RES memory increased more and more, may be memory leak?

ag the sample code like:

 if (QUIC_FAILED(Status = MsQuic->StreamSend(Stream, SendBuffer, 1, QUIC_SEND_FLAG_FIN, SendBuffer))) {
        printf("StreamSend failed, 0x%x!\n", Status);
        free(SendBufferRaw);
        goto Error;
  }

could not find the free when success

**Reply by anrossi** [reply]
The server will grow the internal receive buffer during a transfer, but it should reach a steady state.

In the `ClientStreamCallback` and `ServerStreamCallback`, the `QUIC_STREAM_EVENT_SEND_COMPLETE` event is how MsQuic signals to the application that the send finished and the buffer can be `free`d.

**Comment by kamranattari4192** [comment]
Is QUIC_STREAM_EVENT_IDEAL_SEND_BUFFER_SIZE exposed to .NET API? Is there any sample code that uses this?
